### PR TITLE
Introduce FW boot states

### PR DESCRIPTION
**Pull request by ranj063**
ASoC: SOF: Introduce state machine for FW boot
    
This patch adds the state machine for FW boot that is used to track the different stages of FW boot. This is helpful in keeping track of the current FW boot status that will be used to determine the steps to be performed when the system suspends.

### CODE REVIEW

File: sound/soc/sof/pm.c [hunk 1]
```diff
@@ -56,6 +56,10 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 	if (!sof_ops(sdev)->resume || !sof_ops(sdev)->runtime_resume)
 		return 0;
 
+	/* Nothing to do if firmware never booted */
```
Comment (plbossart) [inline sound/soc/sof/pm.c]: that doesn't even begin to describe the issue...

Also I am not sure if this is right. don't we boot the DSP on resume?

File: sound/soc/sof/pm.c [hunk 2]
```diff
@@ -128,6 +132,10 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 	if (!sof_ops(sdev)->suspend)
 		return 0;
 
+	/* Nothing to do if firmware never booted */
+	if (!sdev->boot_complete)
+		return 0
```
Comment (plbossart) [inline sound/soc/sof/pm.c]: well here we may still want to turn the cores off, no?
maybe this should skip to power down instead of returning?

Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart hmm I missed a big point here. I just realized that we set sdev->boot_complete even if the firmware boot fails. Let me debug this a bit more

**Comment by ranj063** [comment]
@plbossart we set the boot_complete to true even if the firmware fails to boot. Your patch f7f346d588fa0b moved the boot_complete= true to after dumping the data. I wonder why we need to set boot_complete to true at all if the firmware didnt boot? Do you have any context for this? I just cant to recollect it

**Comment by plbossart** [comment]
> @plbossart we set the boot_complete to true even if the firmware fails to boot. Your patch [f7f346d](https://github.com/thesofproject/linux/commit/f7f346d588fa0b0cd91db1f166f5fd598585f64f) moved the boot_complete= true to after dumping the data. I wonder why we need to set boot_complete to true at all if the firmware didnt boot? Do you have any context for this? I just cant to recollect it

I just moved stuff around to fix the error.

I think @kv2019i is da man IIRC

**Comment by ranj063** [comment]
> I think @kv2019i is da man IIRC

@plbossart yes its a69270d8bfe that added this to address a race (likely seen with QEMU)

File: sound/soc/sof/pm.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart if the FW fails to load upon boot, we call snd_sof_remove() and disable the cores. So I guess we dont need to no anything here.

I've updated the PR to use first_boot instead of boot_complete. first_boot is set to false after the first fw boot is successful.

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart yes, we do boot the DSP on resume. But if the firmware boot failed at first boot, it is likely to fail again during resume. So, isnt it better to avoid it?

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @plbossart yes its [a69270d](https://github.com/thesofproject/linux/commit/a69270d8bfeb35fc89d047ea6db803cd75a74f12) that added this to address a race (likely seen with QEMU)

Ack, This was solely to fix bug in the state machine handling FW_READY responses. We had no other logic driven by "boot_complete" back then.

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/pm.c]: @ranj063 @plbossart Hmm, this is starting to get hard to follow and my previous patch that misused "boot_complete" is contributing to the overall difficulty. 

How about if we add an explicit enum of FW states, FW_NOT_STARTED, FW_FIRST_BOOT, FW_BOOT_OK, FW_BOOT_ERROR  and replace both first_boot and boot_complete with the enum? 

Now the above state names are maybe not the best possible, but these would already make the code more self-documenting. E.g. if state==FW_BOOT_ERROR, we can return directly here, the DSP won't respond anyways. The "if (first_boot)" is much less obvious.

Another thing is the conditions when to not retry boot because earlier errors. Maybe it's ok to skip the boot if the initial fw boot failed. But for other errors (a rare FW error during some use-case flow), I'd be more careful as a power cycle and reload FW may be a succesful recovery method in these cases. But yeah, for initial FW load (until FW_READY), for that retry is unlikely to help.

**Comment by ranj063** [comment]
@kv2019i in theory adding enum for FW boot sounds good but we already have all the information we possibly need to make the call I think. I feel lilke this would be a bit of an overkill. We've got the first_boot and we could possibly even consider the sdev->fw_ready to check if the initial FW boot was successful.

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @kv2019i in theory adding enum for FW boot sounds good but we already have all 
> the information we possibly need to make the call I think. I feel lilke this would be a bit 

That's my concern actually that we have too much boolean flags in sofdev, and then we have adhoc state machines that make decisions based on these.

You got mislead by my use of "boot_complete". It's not very obvious that it gets to be true when FW boot fails.

Similarly, I didn't get your patch at first. With "if (first_boot)" you are actually saying that if are trying to suspend with this flag set, something went wrong and we should ignore the suspend.

In both cases, we are gathering code that is hard to maintain.

It's true first_boot has specific semantics that doesn't map so nicely to an enum. What if we add an enum for the FW boot status (IDLE, BOOTING, BOOK_OK, ERROR). This together with "first_boot", I think we can make the code more readable in all places.

**Comment by ranj063** [comment]
> It's true first_boot has specific semantics that doesn't map so nicely to an enum. What if we add an enum for the FW boot status (IDLE, BOOTING, BOOK_OK, ERROR). This together with "first_boot", I think we can make the code more readable in all places.

@kv2019i OK if you think it makes code more maintainable that way, I will make the changes.

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/pm.c]: @kv2019i agree, time for proper state machines. Thanks for suggesting this.

**Comment by mengdonglin** [comment]
@ranj063 There may be another situation need to be covered:

In SOF https://github.com/thesofproject/sof/issues/2117, for TGL RVP, @mmaka1 noticed both https://github.com/thesofproject/sof/files/3885435/dmesg_130.log and https://github.com/thesofproject/sof/files/3885440/dmesg._205.log report successful pm_ctx_save IPC done while snd_sof_runtime_idle followed by a failure while resetting HD/A controller by the driver:

[ 3682.689883] sof-audio-pci 0000:00:1f.3: snd_sof_runtime_idle
[ 3682.689889] sof-audio-pci 0000:00:1f.3: snd_sof_runtime_suspend
[ 3682.689892] sof-audio-pci 0000:00:1f.3: sof_suspend
[ 3682.690347] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x20140000 successful
[ 3682.747154] sof-audio-pci 0000:00:1f.3: sof_send_pm_ctx_ipc
[ 3682.747210] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[ 3682.747830] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
[ 3682.747863] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1010f0f successful
[ 3682.747877] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1000f0f successful
[ 3682.747881] sof-audio-pci 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask 1
[ 3682.747975] sof-audio-pci 0000:00:1f.3: Debug PCIR: 00000010 at  00000044
[ 3683.018196] sof-audio-pci 0000:00:1f.3: error: failed to reset HDA controller gctl 0x1
[ 3683.018203] sof-audio-pci 0000:00:1f.3: error: failed to reset controller during suspend
[ 3683.018206] sof-audio-pci 0000:00:1f.3: error: failed to power down DSP during suspend -5
[ 3683.018213] sof-audio-pci 0000:00:1f.3: can't suspend (snd_sof_runtime_suspend [snd_sof] returned -5)
[ 3688.139530] PM: suspend entry (s2idle)
...
[ 3688.158845] sof-audio-pci 0000:00:1f.3: snd_sof_suspend
[ 3688.158851] sof-audio-pci 0000:00:1f.3: sof_suspend
[ 3688.158854] sof-audio-pci 0000:00:1f.3: sof_send_pm_ctx_ipc
[ 3688.158860] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[ 3688.662540] sof-audio-pci 0000:00:1f.3: error: ipc timed out for 0x40010000 size 76
[ 3688.662543] sof-audio-pci 0000:00:1f.3: info: preventing DSP entering D3 state to preserve context
[ 3688.662561] sof-audio-pci 0000:00:1f.3: unknown ROM status value 01009701
[ 3688.662877] sof-audio-pci 0000:00:1f.3: error: unexpected fault 0x001d003c trace 0x00000000
[ 3688.662908] sof-audio-pci 0000:00:1f.3: error: hda irq intsts 0x00000000 intlctl 0x00000000 rirb 00
[ 3688.662909] sof-audio-pci 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspis 0x00000000
[ 3688.662936] sof-audio-pci 0000:00:1f.3: error: host status 0x00000000 dsp status 0x00040000 mask 0x00000000
[ 3688.662938] sof-audio-pci 0000:00:1f.3: ctx_save ipc error -110, proceeding with suspend

Since the DSP responded to the first pm_ctx_save, it would NOT be able to handle any subsequent IPC interrupts in this state including another pm_ctx_save attempt from the driver.

**Comment by ranj063** [comment]
@mengdonglin thanks for bringing this to my attention. This is a scenario we've never encountered where runtime suspend fails and then we enter system suspend. 

It looks like in the case of TGL the HDA reset seems to timeout. Can we please check if increasing the reset timeout (HDA_DSP_CTRL_RESET_TIMEOUT) for TGL helps with the issue?

**Comment by keyonjie** [comment]
If the previous suspending fail, there will be no corresponding resuming call, so the next PM ops invoked will be the next suspending.

It looks we need to break down and think of how should we handle to failure on suspending carefully, 

if suspend fails, how should we try to recover the Audio DSP and controller?
1. if release trace trigger stop fail, 
2. if CTX_SAVE IPC return error or timeout, 
3. if DSP cores reset & power down fail,
4. if ctrl_link_reset fail,

How should we handle these kinds of errors, to make it possible to recover the audio? 

For each item, we could add some retry mechanism, and if it keep failing, the extreme thinking may be that let's report warning but keep proceeding the subsequent suspending tasks?

**Comment by ranj063** [comment]
@plbossart @kv2019i updated the PR with FW boot states

File: sound/soc/sof/core.c [hunk 3]
```diff
@@ -104,6 +104,8 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 		return ret;
 	}
 
+	sdev->fw_state = SOF_FW_STATE_BOOTING;
+
```
Comment (plbossart) [inline sound/soc/sof/core.c:108]: we have a slew of error below where the firmware boot did not succeed, but the state remains BOOTING.
Shouldn't the state be updated in the fw_err/ipc_err labels?

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -298,6 +298,14 @@ struct snd_sof_ipc_msg {
 	bool ipc_complete;
 };
 
+enum snd_sof_fw_state {
+	SOF_FW_STATE_IDLE = 0,
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: BOOT_NOT_STARTED?

I don't know of a single word synonym for 'not started'

File: sound/soc/sof/loader.c [hunk 5]
```diff
@@ -542,18 +541,21 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 	init_core_mask = ret;
 
 	/* now wait for the DSP to boot */
-	ret = wait_event_timeout(sdev->boot_wait, sdev->boot_complete,
+	ret = wait_event_timeout(sdev->boot_wait,
+				 sdev->fw_state != SOF_FW_STATE_BOOTING,
```
Comment (plbossart) [inline sound/soc/sof/loader.c]: this is odd, shouldn't this more fw_state == STATE_BOOT_OK?

File: sound/soc/sof/pm.c [hunk 6]
```diff
@@ -70,6 +70,8 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 		return ret;
 	}
 
+	sdev->fw_state = SOF_FW_STATE_BOOTING;
+
```
Comment (plbossart) [inline sound/soc/sof/pm.c:74]: same if there are errors the state should transition from  BOOTING to FAILED?

File: sound/soc/sof/core.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/core.c:108]: @plbossart maybe its not obvious but when the FW boot fails, the state gets changed to FAILED during fw run.

File: sound/soc/sof/loader.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/loader.c]: @plbossart this is the part that confused me too but we have a case where the FW booted but there's an ABI mismatch and I added a NOT_READY state for this. This is why I only check for != BOOTING

File: sound/soc/sof/core.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/core.c:108]: what about dbg_err, ipc_err, fw_load_err?

File: sound/soc/sof/loader.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/loader.c]: you should make this explicit with a comment then. It looks like a bug when it's a feature :-)

File: sound/soc/sof/core.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/core.c:108]: dbg_err occurs when we fail to init debugfs
ipc_err occurs when ipc fails to initialize
fw_load error is when fw loading fails.

All of these errors happen before the FW is even attempted to boot. In fact dbg and ipc init only occur after the power cycle and not after a system/runtime resume. So I am not sure if it is helpful to add intermediate state for these. 

But if you think we should, then its not really the FW state but rather the SOF device probe state which would make more sense. But I am not really sure if we really need a state to distinguish these

Comment (plbossart) [reply sound/soc/sof/core.c:108]: why not move this initialization before fw_run? I don't see the point of setting a state that's not really true. In your own words "All of these errors happen before the FW is even attempted to boot."

File: sound/soc/sof/core.c [hunk 7]
```diff
@@ -104,6 +104,8 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 		return ret;
```
Comment (kv2019i) [inline sound/soc/sof/core.c:104]: The commit message could be improved a bit. I think you are adding "a state machine" and not "the state machine". :) I'd reword the start as: "Add an explicit state machine to track stages of firmware boot. This reduce number of distinct state tracking variables and is helpful e.g. when determining the actions to perform at system suspend".

File: sound/soc/sof/loader.c [hunk 8]
```diff
@@ -542,18 +541,21 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
...
 				 msecs_to_jiffies(sdev->boot_timeout));
 	if (ret == 0) {
 		dev_err(sdev->dev, "error: firmware boot failure\n");
 		snd_sof_dsp_dbg_dump(sdev, SOF_DBG_REGS | SOF_DBG_MBOX |
 			SOF_DBG_TEXT | SOF_DBG_PCI);
-		/* after this point FW_READY msg should be ignored */
-		sdev->boot_complete = true;
+		sdev->fw_state = SOF_FW_STATE_BOOT_FAILED;
```
Comment (kv2019i) [inline sound/soc/sof/loader.c:559]: Great, this looks much better than the old (my) code! :)

File: sound/soc/sof/core.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/core.c:108]: Yup, if e.g. ipc_err occurs, we leave FW state as BOOTING and return from the function. If we set state to BOOTING only before run_firmware(), then the state is always accurate. Either boot succeeds or fails. I wouldn't add more states though, seems we don't have real need for them and the proposed states are sufficient.

Comment (ranj063) [reply sound/soc/sof/core.c:108]: @kv2019i @plbossart the reason I set the state to BOOTING right after the DSP probe completes is that if I dont do it here and either debug init or ipc init or fw loading fail, we're never call dsp_remove when suspending the PCI device.

Comment (plbossart) [reply sound/soc/sof/core.c:108]: @ranj063 seems like you have a missing state then? It's really odd to set the state to BOOTING when you are not starting the boot part...

Comment (ranj063) [reply sound/soc/sof/core.c:108]: @plbossart yes indeed. I will refine the states in the next version
